beater/telemetry.go

```diff
@@ -0,0 +1,88 @@
+	"github.com/elastic/beats/v7/libbeat/monitoring"
+)
+
+var apmRegistry = monitoring.GetNamespace("state").GetRegistry().NewRegistry("apm-server")
+
+type configTelemetry struct {
+	rumEnabled                *monitoring.Bool
+	apiKeysEnabled            *monitoring.Bool
+	kibanaEnabled             *monitoring.Bool
+	pipelinesEnabled          *monitoring.Bool
+	pipelinesOverwrite        *monitoring.Bool
+	setupTemplateEnabled      *monitoring.Bool
+	setupTemplateOverwrite    *monitoring.Bool
+	setupTemplateAppendFields *monitoring.Bool
+	ilmEnabled                *monitoring.Bool
+	ilmSetupEnabled           *monitoring.Bool
+	ilmSetupOverwrite         *monitoring.Bool
+	ilmSetupRequirePolicy     *monitoring.Bool
+	jaegerGRPCEnabled         *monitoring.Bool
+	jaegerHTTPEnabled         *monitoring.Bool
+	sslEnabled                *monitoring.Bool
+}
+
+var configMonitors = &configTelemetry{
+	rumEnabled:                monitoring.NewBool(apmRegistry, "rum.enabled"),
+	apiKeysEnabled:            monitoring.NewBool(apmRegistry, "api_key.enabled"),
+	kibanaEnabled:             monitoring.NewBool(apmRegistry, "kibana.enabled"),
+	pipelinesEnabled:          monitoring.NewBool(apmRegistry, "register.ingest.pipeline.enabled"),
+	pipelinesOverwrite:        monitoring.NewBool(apmRegistry, "register.ingest.pipeline.overwrite"),
+	setupTemplateEnabled:      monitoring.NewBool(apmRegistry, "setup.template.enabled"),
+	setupTemplateOverwrite:    monitoring.NewBool(apmRegistry, "setup.template.overwrite"),
+	setupTemplateAppendFields: monitoring.NewBool(apmRegistry, "setup.template.append_files"),
+	ilmEnabled:                monitoring.NewBool(apmRegistry, "ilm.enabled"),
+	ilmSetupEnabled:           monitoring.NewBool(apmRegistry, "ilm.setup.enabled"),
+	ilmSetupOverwrite:         monitoring.NewBool(apmRegistry, "ilm.setup.overwrite"),
+	ilmSetupRequirePolicy:     monitoring.NewBool(apmRegistry, "ilm.setup.require.policy"),
+	jaegerGRPCEnabled:         monitoring.NewBool(apmRegistry, "jaeger.grpc.enabled"),
+	jaegerHTTPEnabled:         monitoring.NewBool(apmRegistry, "jaeger.http.enabled"),
+	sslEnabled:                monitoring.NewBool(apmRegistry, "ssl.enabled"),
+}
+
+func recordConfigs(info beat.Info, apmCfg *config.Config, rootCfg *common.Config) {
+	configMonitors.rumEnabled.Set(apmCfg.RumConfig.IsEnabled())
+	configMonitors.apiKeysEnabled.Set(apmCfg.APIKeyConfig.IsEnabled())
+	configMonitors.kibanaEnabled.Set(apmCfg.Kibana.Enabled)
+	configMonitors.jaegerHTTPEnabled.Set(apmCfg.JaegerConfig.HTTP.Enabled)
+	configMonitors.jaegerGRPCEnabled.Set(apmCfg.JaegerConfig.GRPC.Enabled)
+	configMonitors.sslEnabled.Set(apmCfg.TLS.IsEnabled())
+	configMonitors.pipelinesEnabled.Set(apmCfg.Register.Ingest.Pipeline.IsEnabled())
+	configMonitors.pipelinesOverwrite.Set(apmCfg.Register.Ingest.Pipeline.ShouldOverwrite())
+	indexManagementCfg, err := idxmgmt.NewIndexManagementConfig(info, rootCfg)
+	if err != nil {
+		return
```

Comment:
Sorry, I missed this before: please return the error and have the caller check it. If we get an error here something has gone badly wrong, and we should make some noise about it.

Reply:
> we should make some noise about it.

What do you mean with "noise"?
All other errors returned to `beater` result in the process stopped. Is telemetry recording critical? If so, why? It not, why do you want to stop the process on a non critical error?

If you want to propagate it, what other subroutines should be aborted because of telemetry not working?
If you just want to log it, why it needs to be returned? More so: that error is checked elsewhere (at the point where is actually critical), so in best case the error would be logged twice. 
With current logic the process will stop before even getting to that, so it would be dead code.

Reply:
I agree that the error should not be swallowed here, maybe logging it would be good enough though. How this error is handled here should not depend on whether or not some completely unrelated code part experiencing the same error would already stop the whole APM Server; it's completely independent from a logical point of view. Such implicit dependencies are dangerous (usually no tests would fail when anything changes and it wouldn't be anything obvious to change telemetry if some general program logic order changes).

Reply:
@simitt thank you, you explained my concern perfectly.

Silently swallowing errors is never OK in production code. Code changes, and assumptions made at the time consequently change.

@jalvz I see a few ways forward:

 - Bubble up the error. That _should_ never happen, but if it does happen (e.g. because of code changes) we'll find out about it quickly and have an opportunity to fix it.
 - Log the error and continue. Again, should never happen. If it does, maybe we'll find out about it if we see errors in the logs.
 - Refactor the code to eliminate the possibility of an error. Ideally we shouldn't be parsing the config twice, which is the cause of the issue here. Any errors here are not (currently) caused by telemetry failing, but due to config validation. If we get here and config is invalid, there's a logic error in the code that needs fixing.

Removing the possibility of an error would be ideal, but it doesn't look at all straightforward due to libbeat taking control of setting up index management.